In [1]:
#from lib.project_5 import load_data_from_database, make_data_dict, general_model, general_transformer

# Step 1 - Benchmarking

**NOTE: EACH OF THESE SHOULD BE WRITTEN SOLELY WITH REGARD TO STEP 1 - BENCHMARKING**

### Domain and Data

**TODO:** Write a simple statement about the domain of your problem and the dataset upon which you will be working. 
Prepared for the Neural Information Processing Symposium 2003 Feature Extraction Workshop

http://clopinet.com/isabelle/Projects/NIPS2003

### Data 

MADELON is an artificial dataset, which was part of the NIPS 2003 feature selection challenge. This is a two-class classification problem with continuous input variables. The difficulty is that the problem is multivariate and highly non-linear.

MADELON is an artificial dataset containing data points grouped in 32 clusters placed on the vertices of a five dimensional hypercube and randomly labeled +1 or -1. The five dimensions constitute 5 informative features. 15 linear combinations of those features were added to form a set of 20 (redundant) informative features. Based on those 20 features one must separate the examples into the 2 classes (corresponding to the +-1 labels). We added a number of distractor feature called 'probes' having no predictive power. The order of the features and patterns were randomized. 

### Problem Statement

**TODO:** Write a simple problem statement with regard to benchmarking your work only.
The NIPS 2003 challenge in feature selection is to find feature selection algorithms that significantly outperform methods using all features in performing a binary classification task.

### Solution Statement

**TODO:** Write a simple solution statement with regard to benchmarking your work only.
We will develop a binary classification model and attempt to augment its performance using automatic feature selection techniques.

### Metric

**TODO**: Write a statement about the metric you will be using. This section is global as it will be the metric you will use throughout this project. Provide a brief justification for choosing this metric.

We will examine all of the metrics that we have used so far - Accuracy, Precision, Recall, and F1-Score, as well as look at a Confusion Matrix. For comparing models, we will use Accuracy.

### Benchmark

**TODO**: Write a statement explaining that this is the process by which you identify a benchmark for your project.
We will use as a benchmark our mean accuracy across five random train test splits using a K Nearest Neighbors model with an optimal value for number of `n_neighbors`. This model had a 73.8% accuracy.


## Implementation

Implement the following code pipeline using the functions you write in `lib/project_5.py`.

<img src="assets/benchmarking.png" width="600px">

In [1]:
from sqlalchemy import create_engine
import pandas as pd
#from os import chdir; chdir('..')
from os import chdir; chdir('./lib')
from sklearn.preprocessing import StandardScaler
from project_5 import load_data_from_database, make_data_dict,general_transformer, general_model
from sklearn import linear_model
#from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression,Lasso
from sklearn import metrics
from sklearn.feature_selection import SelectKBest, f_regression, f_classif
from sklearn.pipeline import Pipeline
from sklearn.pipeline import make_pipeline
from sklearn.neighbors import KNeighborsClassifier


In [41]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [38]:
!pwd

/src/dsi/dsi-workspace/project-05/lib


In [ ]:
pd.read_sql()

In [3]:
madelon_df = load_data_from_database()
madelon_df.head(5)

index  feat_000  feat_001  feat_002  feat_003  feat_004  feat_005  \
0      0       485       477       537       479       452       471   
1      1       483       458       460       487       587       475   
2      2       487       542       499       468       448       471   
3      3       480       491       510       485       495       472   
4      4       484       502       528       489       466       481   

   feat_006  feat_007  feat_008  ...    feat_491  feat_492  feat_493  \
0       491       476       475  ...         481       477       485   
1       526       479       485  ...         478       487       338   
2       442       478       480  ...         481       492       650   
3       417       474       502  ...         480       474       572   
4       402       478       487  ...         479       452       435   

   feat_494  feat_495  feat_496  feat_497  feat_498  feat_499  label  
0       511       485       481       479       475       496     -1  
1       513       486       483       492       510       517     -1  
2       506       501       480       489       499       498     -1  
3       454       469       475       482       494       461      1  
4       486       508       481       504       495       511      1  

[5 rows x 502 columns]

In [4]:
madelon_df.shape

(2000, 502)

In [5]:
madelon_df.describe().T

count      mean         std    min     25%    50%      75%     max
index     2000.0  999.5000  577.494589    0.0  499.75  999.5  1499.25  1999.0
feat_000  2000.0  481.7225    6.421769  462.0  477.00  482.0   486.00   503.0
feat_001  2000.0  483.4525   30.186294  381.0  464.00  483.0   503.00   600.0
feat_002  2000.0  510.1660   38.899165  370.0  485.00  510.5   536.00   654.0
feat_003  2000.0  483.3845    9.059895  453.0  477.00  483.0   490.00   519.0
feat_004  2000.0  501.6125   41.389418  371.0  475.00  500.0   528.00   688.0
feat_005  2000.0  479.2590    6.795956  459.0  475.00  479.0   484.00   505.0
feat_006  2000.0  480.1095   40.575925  334.0  452.75  480.0   506.25   611.0
feat_007  2000.0  476.5650    1.384461  471.0  476.00  477.0   477.00   481.0
feat_008  2000.0  486.7935   15.043836  430.0  477.00  487.0   496.25   536.0
feat_009  2000.0  478.7890    7.190920  455.0  474.00  479.0   484.00   503.0
feat_010  2000.0  486.5175   36.556831  354.0  462.00  486.0   511.00   620.0
feat_011  2000.0  490.6535   28.198404  389.0  472.00  491.0   508.25   603.0
feat_012  2000.0  478.1645   34.169434  347.0  455.00  478.0   502.00   593.0
feat_013  2000.0  481.5245   42.617385  352.0  454.00  482.0   510.00   632.0
feat_014  2000.0  485.0385   12.135224  444.0  477.00  485.0   493.00   523.0
feat_015  2000.0  479.4235   21.984604  410.0  464.00  479.0   494.00   551.0
feat_016  2000.0  479.2870   13.483602  433.0  470.00  479.0   489.00   523.0
feat_017  2000.0  494.9475   30.339495  377.0  475.00  495.0   515.00   601.0
feat_018  2000.0  484.1730   23.369449  408.0  468.00  484.0   500.00   566.0
feat_019  2000.0  477.4270    9.957170  441.0  470.00  477.5   484.00   513.0
feat_020  2000.0  484.7145   12.724305  426.0  476.75  485.0   493.00   526.0
feat_021  2000.0  494.2390   22.344295  412.0  479.00  494.0   509.00   571.0
feat_022  2000.0  476.5730    6.840326  456.0  472.00  477.0   481.00   499.0
feat_023  2000.0  479.2605    9.641946  438.0  473.00  479.0   486.00   516.0
feat_024  2000.0  499.4075   44.247580  343.0  469.00  500.0   529.00   653.0
feat_025  2000.0  504.7190   28.529686  416.0  485.00  504.0   524.00   597.0
feat_026  2000.0  484.7835   21.573278  421.0  471.00  485.0   499.00   573.0
feat_027  2000.0  482.0995   11.934899  441.0  474.00  482.0   490.00   528.0
feat_028  2000.0  480.5975   14.703897  438.0  468.00  481.0   493.00   520.0
...          ...       ...         ...    ...     ...    ...      ...     ...
feat_471  2000.0  478.7120    4.541957  464.0  476.00  479.0   482.00   493.0
feat_472  2000.0  480.3980   44.456189  347.0  451.00  480.0   512.00   621.0
feat_473  2000.0  477.6405    1.787696  472.0  476.00  478.0   479.00   484.0
feat_474  2000.0  479.4110    6.381570  455.0  475.00  479.0   484.00   506.0
feat_475  2000.0  486.3710   74.518547  276.0  424.00  485.5   547.25   680.0
feat_476  2000.0  481.3660   15.903317  428.0  471.00  482.0   492.00   535.0
feat_477  2000.0  487.7350   17.023180  423.0  476.00  488.0   499.00   548.0
feat_478  2000.0  483.6760   37.987953  347.0  458.00  483.0   509.00   630.0
feat_479  2000.0  486.6320   11.695810  450.0  479.00  487.0   495.00   528.0
feat_480  2000.0  485.4965   19.957333  418.0  472.00  486.0   499.00   554.0
feat_481  2000.0  478.4575   27.820985  391.0  460.00  478.0   497.00   587.0
feat_482  2000.0  477.3605    9.774779  439.0  471.00  478.0   484.00   509.0
feat_483  2000.0  497.9700   34.666092  375.0  475.00  497.0   521.00   623.0
feat_484  2000.0  501.8770   35.628971  376.0  478.00  501.0   526.25   619.0
feat_485  2000.0  482.7790   11.275936  444.0  475.00  483.0   490.00   523.0
feat_486  2000.0  478.7155   10.149387  444.0  472.00  479.0   485.00   512.0
feat_487  2000.0  482.5695   34.051431  355.0  459.00  482.0   505.00   618.0
feat_488  2000.0  489.9115   22.729650  411.0  475.00  490.0   505.00   606.0
feat_489  2000.0  479.6215    8.864622  449.0  474.00  480.0   485.00   513.0
feat_490  2000.0  482.4585   20.8

In [6]:
madelon_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Columns: 502 entries, index to label
dtypes: int64(502)
memory usage: 7.7 MB


**Check for null values in all columns.**

In [7]:
[madelon_df[pd.isnull(madelon_df[col])].count() for col in madelon_df.columns] 

[index       0
 feat_000    0
 feat_001    0
 feat_002    0
 feat_003    0
 feat_004    0
 feat_005    0
 feat_006    0
 feat_007    0
 feat_008    0
 feat_009    0
 feat_010    0
 feat_011    0
 feat_012    0
 feat_013    0
 feat_014    0
 feat_015    0
 feat_016    0
 feat_017    0
 feat_018    0
 feat_019    0
 feat_020    0
 feat_021    0
 feat_022    0
 feat_023    0
 feat_024    0
 feat_025    0
 feat_026    0
 feat_027    0
 feat_028    0
            ..
 feat_471    0
 feat_472    0
 feat_473    0
 feat_474    0
 feat_475    0
 feat_476    0
 feat_477    0
 feat_478    0
 feat_479    0
 feat_480    0
 feat_481    0
 feat_482    0
 feat_483    0
 feat_484    0
 feat_485    0
 feat_486    0
 feat_487    0
 feat_488    0
 feat_489    0
 feat_490    0
 feat_491    0
 feat_492    0
 feat_493    0
 feat_494    0
 feat_495    0
 feat_496    0
 feat_497    0
 feat_498    0
 feat_499    0
 label       0
 dtype: int64, index       0
 feat_000    0
 feat_001    0
 feat_002    0
 feat_003  

In [8]:
madelon_df["label"].unique()

array([-1,  1])

** Calculate the Baseline Accuracy **

**`baseline_accuracy = majority_class_N / total_population`**

In [9]:
madelon_df.groupby(by="label")["index"].count()

label
-1    1000
 1    1000
Name: index, dtype: int64

In [10]:
baseline_accuracy = 1000*1./madelon_df["label"].count()
baseline_accuracy

0.5

**Baseline Accuracy is Set to 50%**

## Define X and Y

In [11]:
# looking at the mean of Index column and its name it seems it is a counter and does not contribute towards 
# the study, so we will remove that column from study
del madelon_df["index"]
# Define X and y
y = madelon_df["label"]
X = madelon_df.drop("label",axis=1)

In [12]:
data_dict= make_data_dict(X,y,0.5,random_state=42)

In [13]:
data_dict['X_train'].head()

feat_000  feat_001  feat_002  feat_003  feat_004  feat_005  feat_006  \
440       473       470       511       480       555       480       475   
573       481       503       527       479       529       479       447   
946       477       445       499       497       434       475       481   
997       491       468       495       484       548       479       533   
503       469       489       494       492       445       484       484   

     feat_007  feat_008  feat_009    ...     feat_490  feat_491  feat_492  \
440       476       493       485    ...          482       478       469   
573       475       484       469    ...          490       481       474   
946       478       500       482    ...          523       477       482   
997       478       465       464    ...          494       474       504   
503       477       474       481    ...          481       483       510   

     feat_493  feat_494  feat_495  feat_496  feat_497  feat_498  feat_499  
440       505       487       529       481       482       524       491  
573       672       461       537       469       496       572       476  
946       788       488       479       475       495       501       474  
997       473       489       483       476       494       565       457  
503       613       486       528       485       504       510       511  

[5 rows x 500 columns]

In [14]:
data_dict = general_transformer(StandardScaler(),data_dict)
data_dict

{'X_test': array([[ 1.12156758, -0.51026631,  0.33005212, ..., -0.52713324,
         -0.07137678,  0.47371054],
        [-0.3047583 , -0.67637944,  1.08365312, ..., -0.01450019,
         -0.07137678,  0.47371054],
        [ 0.17068366, -1.34083196, -0.64962918, ...,  1.66986554,
          0.79543799, -0.22014492],
        ..., 
        [ 0.17068366, -0.77604732, -0.09698845, ..., -1.47916605,
          0.30785468,  0.04968776],
        [ 0.80460628,  2.34687951,  0.48077232, ...,  1.96279872,
         -0.28808047,  0.82063828],
        [ 0.48764497, -0.04514955, -2.15683118, ..., -0.74683312,
         -2.02171001,  1.01337591]]),
 'X_train': array([[-1.41412287, -0.44382106,  0.02861172, ..., -0.08773348,
          0.44329449,  0.04968776],
        [-0.14627764,  0.65252559,  0.43053225, ...,  0.93753262,
          1.74351665, -0.52852513],
        [-0.78020026, -1.2743867 , -0.27282868, ...,  0.86429932,
         -0.17972863, -0.60562018],
        ..., 
        [-0.14627764,  0.287076

In [15]:
data_dict.keys()

['X_test', 'processes', 'X_train', 'y_train', 'y_test']

In [16]:
transformer = data_dict["processes"][0]

In [17]:
data_dict.keys()

['X_test', 'processes', 'X_train', 'y_train', 'y_test']

**Below model is EDA Purposes only in which we are going to leverage sklearn:**

1. We leverage standardized and transformed train data from tranformed process. 
2. Fit the train feature and target set 
4. Create a Logistic Regression with l2 i.e. Ridge with large C value to minimize regularization.  
6. Evalute the score.
7. Review the coefficients
8. Create a mapping of coefficient and features
9. Write analysis on my finding

In [18]:
# run preliminary logistic regression for EDA purpose
model = linear_model.LogisticRegression(penalty = 'l2', C=1E10)
data_dict = general_model(model, data_dict)
#model.fit(data_dict["X_train"], data_dict["y_train"])
data_dict.keys()

['X_test',
 'processes',
 'X_train',
 'test_score',
 'train_score',
 'y_train',
 'y_test']

In [20]:
model = data_dict["processes"][1]

In [34]:
print "Train Score:", data_dict["train_score"], "Test Score:", data_dict["test_score"]

Train Score: 1.0 Test Score: 0.544


In [37]:
feature_coef = []
for i,k in enumerate(X.columns):
    feature_coef.append([k, model.coef_[0][i]])
df_coef=pd.DataFrame(feature_coef)
df_coef.columns = ["feature","coef_"]

df_coef.sort_values(["coef_"],ascending=False)


feature      coef_
475  feat_475  34.224080
453  feat_453  29.932129
46   feat_046  29.267086
317  feat_317  24.742018
136  feat_136  22.464534
251  feat_251  20.894102
445  feat_445  20.791725
151  feat_151  20.580548
455  feat_455  20.216010
116  feat_116  19.432239
56   feat_056  18.990340
372  feat_372  18.230913
41   feat_041  17.981450
196  feat_196  17.695247
424  feat_424  17.311354
340  feat_340  16.453879
186  feat_186  16.094626
446  feat_446  15.871906
336  feat_336  15.871505
454  feat_454  15.848546
320  feat_320  15.659518
98   feat_098  15.268816
184  feat_184  14.944141
349  feat_349  14.569727
211  feat_211  14.469743
244  feat_244  14.416206
73   feat_073  14.137550
471  feat_471  14.063048
75   feat_075  13.897754
378  feat_378  13.332955
..        ...        ...
334  feat_334 -13.335155
459  feat_459 -13.992923
413  feat_413 -14.105048
436  feat_436 -14.288182
235  feat_235 -14.821088
89   feat_089 -14.952522
297  feat_297 -14.983977
447  feat_447 -15.239271
423  feat_423 -15.510017
63   feat_063 -15.767075
243  feat_243 -15.801745
198  feat_198 -16.215252
127  feat_127 -16.450219
208  feat_208 -16.578771
161  feat_161 -16.648343
93   feat_093 -16.803694
31   feat_031 -16.829499
462  feat_462 -17.111785
315  feat_315 -17.514917
491  feat_491 -17.740405
278  feat_278 -17.838318
130  feat_130 -18.773787
441  feat_441 -19.462824
296  feat_296 -19.601788
168  feat_168 -20.396778
60   feat_060 -21.106476
495  feat_495 -21.338762
5    feat_005 -21.687148
442  feat_442 -24.560974
153  feat_153 -29.341438

[500 rows x 2 columns]

With the naive LogisticRegression it appears that model worked very well on train set however performed poorly on the test data set. ** The train score was perfectly 1.0 and test data score dcid poorly 0.544 **. This suggest we need to further tune the model with more penaly or even performing LogisticRegression using Lasso method. 